packages/backend/src/managers/podmanConnection.ts

```diff
@@ -16,94 +16,163 @@
-  }
+  protected refreshProviders(): void {
+    // clear all providers
+    this.#providers.clear();
 
-  // startupSubscribe registers f to be executed when a podman container provider
-  // registers, or immediately if already registered
-  startupSubscribe(f: startupHandle): void {
-    if (this.#firstFound) {
-      f();
-    } else {
-      this.#toExecuteAtStartup.push(f);
-    }
+    const providers = provider.getContainerConnections();
+
+    // register the podman container connection
+    providers
+      .filter(({ connection }) => connection.type === 'podman')
+      .forEach(({ providerId, connection }) => {
+        this.#providers.set(providerId, [connection, ...(this.#providers.get(providerId) ?? [])]);
+      });
+
+    // notify
+    this.notify();
   }
 
-  listenMachine() {
-    provider.onDidUpdateContainerConnection((e: UpdateContainerConnectionEvent) => {
-      if (e.connection.type !== 'podman') {
-        return;
-      }
-      if (e.status === 'stopped') {
-        for (const f of this.#toExecuteAtMachineStop) {
-          f();
+  private listen() {
+    // capture unregister event
+    this.#disposables.push(
+      provider.onDidUnregisterContainerConnection(() => {
+        this.refreshProviders();
+        this._onPodmanConnectionEvent.fire({
+          status: 'unregister',
+        });
+      }),
+    );
+
+    this.#disposables.push(
+      provider.onDidRegisterContainerConnection(({ providerId, connection }: RegisterContainerConnectionEvent) => {
+        if (connection.type !== 'podman') {
+          return;
         }
-      } else if (e.status === 'started') {
-        for (const f of this.#toExecuteAtMachineStart) {
-          f();
+
+        // update connection
+        this.#providers.set(providerId, [connection, ...(this.#providers.get(providerId) ?? [])]);
+        this.notify();
+        this._onPodmanConnectionEvent.fire({
+          status: 'register',
+        });
+      }),
+    );
+
+    this.#disposables.push(
+      provider.onDidUpdateContainerConnection(({ status }: UpdateContainerConnectionEvent) => {
+        switch (status) {
+          case 'started':
+          case 'stopped':
+            this._onPodmanConnectionEvent.fire({
+              status: status,
+            });
+            this.notify();
+            break;
+          default:
+            break;
         }
-      }
-    });
+      }),
+    );
+
+    this.#disposables.push(
+      provider.onDidUpdateProvider(() => {
+        this.refreshProviders();
+      }),
+    );
+  }
+
+  protected parseVMType(vmtype: string | undefined): VMType {
+    switch (vmtype) {
+      // mac
+      case VMType.APPLEHV:
+        return VMType.APPLEHV;
+      case VMType.QEMU:
+        return VMType.QEMU;
+      case VMType.LIBKRUN:
+        return VMType.LIBKRUN;
+      // windows
+      case VMType.HYPERV:
+        return VMType.HYPERV;
+      case VMType.WSL:
+        return VMType.WSL;
+      default:
+        return VMType.UNKNOWN;
```

Comment:
here it looks we're doing a lot of stuff

maybe simpler stuff like that will take into account new properties added to the VMType enum later
```ts
  const type = VMType[vmtype.toUpperCase()];
  if (type === undefined) {
    return VMType. UNKNOWN;
  }
return type
```

Reply:
I will give it a try thanks for the shortening